notify/smtp.sh

```diff
@@ -2,14 +2,330 @@
+  _debug "SMTP_BIN" "$SMTP_BIN"
+
+  _debug "_SMTP_SUBJECT" "$_SMTP_SUBJECT"
+  _debug "_SMTP_CONTENT" "$_SMTP_CONTENT"
+
+  # Validate config and apply defaults:
+  # _SMTP_* variables are the resolved (with defaults) versions of SMTP_*.
+  # (The _SMTP_* versions will not be stored in account conf.)
+
+  if [ -n "$SMTP_BIN" ] && ! _exists "$SMTP_BIN"; then
+    _err "SMTP_BIN '$SMTP_BIN' does not exist."
+    return 1
+  fi
+  _SMTP_BIN="$SMTP_BIN"
+  if [ -z "$_SMTP_BIN" ]; then
+    # Look for a command that can communicate with an SMTP server.
+    # (Please don't add sendmail, ssmtp, mutt, mail, or msmtp here.
+    # Those are already handled by the "mail" notify hook.)
+    for cmd in curl python3 python2.7 python pypy3 pypy; do
+      if _exists "$cmd"; then
+        _SMTP_BIN="$cmd"
+        break
+      fi
+    done
+    if [ -z "$_SMTP_BIN" ]; then
+      _err "The smtp notify-hook requires curl or Python, but can't find any."
+      _err 'If you have one of them, define SMTP_BIN="/path/to/curl_or_python".'
+      _err 'Otherwise, see if you can use the "mail" notify-hook instead.'
+      return 1
+    fi
+    _debug "_SMTP_BIN" "$_SMTP_BIN"
+  fi
+
+  if [ -z "$SMTP_FROM" ]; then
+    _err "You must define SMTP_FROM as the sender email address."
+    return 1
+  fi
+  _SMTP_FROM="$SMTP_FROM"
+
+  if [ -z "$SMTP_TO" ]; then
+    _err "You must define SMTP_TO as the recipient email address."
+    return 1
+  fi
+  _SMTP_TO="$SMTP_TO"
+
+  if [ -z "$SMTP_HOST" ]; then
+    _err "You must define SMTP_HOST as the SMTP server hostname."
+    return 1
+  fi
+  _SMTP_HOST="$SMTP_HOST"
+
+  _SMTP_SECURE="${SMTP_SECURE:-none}"
+  case "$_SMTP_SECURE" in
+  "none") smtp_default_port="25" ;;
+  "ssl") smtp_default_port="465" ;;
+  "tls") smtp_default_port="587" ;;
+  *)
+    _err "Invalid SMTP_SECURE='$SMTP_SECURE'. It must be 'ssl', 'tls' or 'none'."
+    return 1
+    ;;
+  esac
+
+  _SMTP_PORT="${SMTP_PORT:-$smtp_default_port}"
+  if [ -z "$SMTP_PORT" ]; then
+    _debug "_SMTP_PORT" "$_SMTP_PORT"
+  fi
+
+  _SMTP_USERNAME="$SMTP_USERNAME"
+  _SMTP_PASSWORD="$SMTP_PASSWORD"
+  _SMTP_TIMEOUT="${SMTP_TIMEOUT:-30}"
+
+  # Run with --debug 2 (or above) to echo the transcript of the SMTP session.
+  # Careful: this may include SMTP_PASSWORD in plaintext!
+  if [ "${DEBUG:-$DEBUG_LEVEL_NONE}" -ge "$DEBUG_LEVEL_2" ]; then
+    _SMTP_SHOW_TRANSCRIPT="True"
+  else
+    _SMTP_SHOW_TRANSCRIPT=""
+  fi
+
+  # Send the message:
+  case "$(basename "$_SMTP_BIN")" in
+  curl) _smtp_send=_smtp_send_curl ;;
+  py*) _smtp_send=_smtp_send_python ;;
+  *)
+    _err "Can't figure out how to invoke $_SMTP_BIN."
+    _err "Check your SMTP_BIN setting."
+    return 1
+    ;;
+  esac
+
+  if ! smtp_output="$($_smtp_send)"; then
+    _err "Error sending message with $_SMTP_BIN."
+    if [ -n "$smtp_output" ]; then
+      _err "$smtp_output"
+    fi
+    return 1
+  fi
+
+  # Save config only if send was successful:
+  _saveaccountconf_mutable SMTP_BIN "$SMTP_BIN"
```

Comment:
we don't need to save all these variables.

please save them at the begging, only when the user specifies a non-default value.

Reply:
I initially had coded it that way, calling `_saveaccountconf_mutable()` only for non-default values. But then there was no way to switch back to the default value, other than manually editing `account.conf` to remove the overrides.

It seems like we would need to call `_clearaccountconf_mutable()` if the user has specified the default value. That function doesn't exist. (The current mail notify hook tries to [implement this logic](https://github.com/acmesh-official/acme.sh/blob/2.8.8/notify/mail.sh#L19-L24):

```sh
  MAIL_NOVALIDATE="${MAIL_NOVALIDATE:-$(_readaccountconf_mutable MAIL_NOVALIDATE)}"
  if [ -n "$MAIL_NOVALIDATE" ]; then
    _saveaccountconf_mutable MAIL_NOVALIDATE 1
  else
    _clearaccountconf "MAIL_NOVALIDATE"
  fi
```

... but it calls `_clearaccountconf()`, which I think clears the wrong variable name.)

Should I add `_clearaccountconf_mutable()` to the main `acme.sh`?
```sh
#key
_clearaccountconf_mutable() {
  _clearaccountconf "SAVED_$1"
}
```

Or another option could be to add a "default" parameter to the existing `_saveaccountconf_mutable()`, and make it save the value if non-default, otherwise clear the value. That avoids duplicating the "if non-default save otherwise clear" logic in lots of other places:
```sh
#key  value base64encode default
_saveaccountconf_mutable() {
  if [ "$2" = "$4" ]; then
    _clear_conf "$ACCOUNT_CONF_PATH" "SAVED_$1"
  else
    _save_conf "$ACCOUNT_CONF_PATH" "SAVED_$1" "$2" "$3"
  fi
  #remove later
  _clearaccountconf "$1"
}
```

Reply:
>> Should I add _clearaccountconf_mutable() to the main acme.sh?

Yes, go ahead.

But I think it should clear both names:   `XXX` and `SAVED_XXXX`

what do you think?

Reply:
Agreed, `_clearaccountconf_mutable()` should clear both names. Added to main acme.sh.

I've reworked the code to save the variables at the beginning (as they're validated), and only if they are not the default values, as requested. 

I added a couple of helpers in notify/smtp.sh (at the bottom) to reduce some of the repetitive code around handling default values. Let me know if those belong somewhere else.